server/src/main/java/org/elasticsearch/transport/InboundHandler.java

```diff
@@ -147,54 +145,57 @@ private void messageReceived(InboundMessage message, TcpChannel channel) throws
         }
     }
 
-    private <T extends TransportRequest> void handleRequest(TcpChannel channel, Header header, StreamInput stream, int messageLengthBytes) {
+    private <T extends TransportRequest> void handleRequest(TcpChannel channel, Header header, InboundMessage message) throws IOException {
         final String action = header.getActionName();
         final long requestId = header.getRequestId();
         final Version version = header.getVersion();
-        TransportChannel transportChannel = null;
-        try {
+        if (header.isHandshake()) {
             messageListener.onRequestReceived(requestId, action);
-            if (header.isHandshake()) {
-                // Handshakes are not currently circuit broken
-                transportChannel = new TcpTransportChannel(outboundHandler, channel, action, requestId, version,
-                    circuitBreakerService, 0, header.isCompressed(), header.isHandshake());
+            // Cannot short circuit handshakes
+            assert message.isShortCircuit() == false;
+            final StreamInput stream = namedWriteableStream(message.openOrGetStreamInput());
+            assertRemoteVersion(stream, header.getVersion());
+            try (Releasable breakerRelease = message.takeBreakerReleaseControl()) {
+                final TransportChannel transportChannel = new TcpTransportChannel(outboundHandler, channel, action, requestId, version,
+                    header.isCompressed(), header.isHandshake(), () -> {});
                 handshaker.handleHandshake(transportChannel, requestId, stream);
```

Comment:
If `TransportHandshaker.handleHandshake` throws an exception (e.g. IllegalStateException), that is no longer bubbled up back to the node that initiated the handshake. I'm not sure what this change of behavior entails, but would suggest backing it out of this PR.

Reply:
Made this change.